File: backend/app/models/wiki.py
"""
World Wiki models - the unified narrative backbone.

WikiEntry: Core building block of the World Wiki
WikiChange: Pending AI changes awaiting author approval
WikiCrossReference: Links between wiki entries
"""

from sqlalchemy import Column, String, Text, DateTime, Integer, Float, ForeignKey, JSON
from sqlalchemy.orm import relationship
from datetime import datetime
import uuid
from enum import Enum

from app.database import Base


class WikiEntryType(str, Enum):
    """Types of wiki entries"""
    # Characters
    CHARACTER = "character"
    CHARACTER_ARC = "character_arc"
    CHARACTER_RELATIONSHIP = "character_relationship"

    # Locations
    LOCATION = "location"
    LOCATION_HISTORY = "location_history"

    # World Rules
    MAGIC_SYSTEM = "magic_system"
    WORLD_RULE = "world_rule"
    TECHNOLOGY = "technology"
    CULTURE = "culture"
    RELIGION = "religion"

    # Narrative Elements
    FACTION = "faction"
    ARTIFACT = "artifact"
    CREATURE = "creature"
    EVENT = "event"  # Historical events

    # Meta
    TIMELINE_FACT = "timeline_fact"
    THEME = "theme"


class WikiEntryStatus(str, Enum):
    """Status of wiki entries"""
    DRAFT = "draft"
    PUBLISHED = "published"
    ARCHIVED = "archived"


class WikiChangeType(str, Enum):
    """Types of wiki changes"""
    CREATE = "create"
    UPDATE = "update"
    MERGE = "merge"
    DELETE = "delete"


class WikiChangeStatus(str, Enum):
    """Status of wiki changes in approval queue"""
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"


class WikiReferenceType(str, Enum):
    """Types of cross-references between wiki entries"""
    MENTIONS = "mentions"
    RELATED_TO = "related_to"
    PART_OF = "part_of"
    CONFLICTS_WITH = "conflicts_with"
    DEPENDS_ON = "depends_on"
    CHILD_OF = "child_of"
    ALLY_OF = "ally_of"
    ENEMY_OF = "enemy_of"
    OWNS = "owns"
    MEMBER_OF = "member_of"
    LOCATED_IN = "located_in"

    # Culture-specific reference types
    BORN_IN = "born_in"
    EXILED_FROM = "exiled_from"
    ADOPTED_INTO = "adopted_into"
    LEADER_OF = "leader_of"
    REBEL_AGAINST = "rebel_against"
    WORSHIPS = "worships"
    TRADES_WITH = "trades_with"
    ORIGINATED_IN = "originated_in"
    SACRED_TO = "sacred_to"
    RESENTS = "resents"


class WikiEntry(Base):
    """
    Core wiki entry - the building block of the World Wiki.

    Lives at World/Series level and is shared across manuscripts.
    Can be auto-populated by AI or manually created by author.
    """
    __tablename__ = "wiki_entries"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    world_id = Column(String, ForeignKey("worlds.id"), nullable=False)

    # Entry metadata
    entry_type = Column(String, nullable=False)  # WikiEntryType value
    title = Column(String, nullable=False)
    slug = Column(String, nullable=False)  # URL-friendly: "john-smith"

    # Content (structured JSON for type-specific fields + free-form)
    structured_data = Column(JSON, default=dict)  # Type-specific fields
    content = Column(Text)  # Free-form markdown content
    summary = Column(Text)  # AI-generated summary

    # Image/visual
    image_url = Column(String, nullable=True)
    image_seed = Column(Integer, nullable=True)

    # Hierarchical organization
    parent_id = Column(String, ForeignKey("wiki_entries.id"), nullable=True)

    # Link to Codex entity (if this wiki entry originated from an entity)
    linked_entity_id = Column(String, ForeignKey("entities.id"), nullable=True)

    # Source tracking - which manuscripts/chapters reference this
    source_manuscripts = Column(JSON, default=list)  # List of manuscript IDs
    source_chapters = Column(JSON, default=list)  # List of {manuscript_id, chapter_id, excerpt}

    # Status and confidence
    status = Column(String, default=WikiEntryStatus.DRAFT.value)
    confidence_score = Column(Float, default=1.0)  # How confident is the data (1.0 = manual, <1.0 = AI)
    last_verified_at = Column(DateTime, nullable=True)

    # Metadata
    tags = Column(JSON, default=list)  # User-defined tags
    aliases = Column(JSON, default=list)  # Alternative names for search

    # Tracking
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    created_by = Column(String, default="author")  # "ai" or "author"

    # Relationships
    world = relationship("World", backref="wiki_entries")
    parent = relationship("WikiEntry", remote_side=[id], backref="children")
    linked_entity = relationship("Entity", backref="wiki_entry")

    # Cross-references (outgoing)
    outgoing_references = relationship(
        "WikiCrossReference",
        foreign_keys="WikiCrossReference.source_entry_id",
        back_populates="source_entry",
        cascade="all, delete-orphan"
    )

    # Cross-references (incoming)
    incoming_references = relationship(
        "WikiCrossReference",
        foreign_keys="WikiCrossReference.target_entry_id",
        back_populates="target_entry",
        cascade="all, delete-orphan"
    )

    # Pending changes
    pending_changes = relationship(
        "WikiChange",
        back_populates="wiki_entry",
        cascade="all, delete-orphan"
    )

    def __repr__(self):
        return f"<WikiEntry(id={self.id}, type={self.entry_type}, title='{self.title}')>"


class WikiChange(Base):
    """
    Pending changes from AI that need author approval.

    All AI-suggested updates go through this approval queue
    before being applied to wiki entries.
    """
    __tablename__ = "wiki_changes"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    wiki_entry_id = Column(String, ForeignKey("wiki_entries.id"), nullable=True)  # Null for new entries
    world_id = Column(String, ForeignKey("worlds.id"), nullable=False)

    # Change details
    change_type = Column(String, nullable=False)  # WikiChangeType value
    field_changed = Column(String, nullable=True)  # Which field, or null for full entry
    old_value = Column(JSON, nullable=True)  # Previous value (for updates)
    new_value = Column(JSON, nullable=False)  # Proposed new value

    # For CREATE changes, this holds the full proposed entry
    proposed_entry = Column(JSON, nullable=True)

    # AI reasoning
    reason = Column(Text)  # Why AI suggests this change
    source_text = Column(Text)  # The manuscript text that triggered this
    source_chapter_id = Column(String, ForeignKey("chapters.id"), nullable=True)
    source_manuscript_id = Column(String, ForeignKey("manuscripts.id"), nullable=True)
    confidence = Column(Float, default=0.8)  # AI confidence in this suggestion

    # Status
    status = Column(String, default=WikiChangeStatus.PENDING.value)
    reviewed_at = Column(DateTime, nullable=True)
    reviewer_note = Column(Text, nullable=True)  # Author's note on approval/rejection

    # Priority for sorting in queue
    priority = Column(Integer, default=0)  # Higher = more important

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow)

    # Relationships
    wiki_entry = relationship("WikiEntry", back_populates="pending_changes")
    world = relationship("World", backref="wiki_changes")
    source_chapter = relationship("Chapter", foreign_keys=[source_chapter_id])
    source_manuscript = relationship("Manuscript", foreign_keys=[source_manuscript_id])

    def __repr__(self):
        return f"<WikiChange(id={self.id}, type={self.change_type}, status={self.status})>"


class WikiCrossReference(Base):
    """
    Links between wiki entries.

    Enables rich interconnections like:
    - Character A is an ally of Character B
    - Location X is part of Region Y
    - Artifact Z is owned by Character A
    """
    __tablename__ = "wiki_cross_references"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    source_entry_id = Column(String, ForeignKey("wiki_entries.id"), nullable=False)
    target_entry_id = Column(String, ForeignKey("wiki_entries.id"), nullable=False)

    # Reference type
    reference_type = Column(String, nullable=False)  # WikiReferenceType value

    # Additional context
    context = Column(Text, nullable=True)  # Where/why this reference exists
    context_chapter_id = Column(String, ForeignKey("chapters.id"), nullable=True)

    # Display options
    bidirectional = Column(Integer, default=1)  # Show on both entries? (1=yes, 0=no)
    display_label = Column(String, nullable=True)  # Custom label override

    # Tracking
    created_at = Column(DateTime, default=datetime.utcnow)
    created_by = Column(String, default="author")  # "ai" or "author"

    # Relationships
    source_entry = relationship(
        "WikiEntry",
        foreign_keys=[source_entry_id],
        back_populates="outgoing_references"
    )
    target_entry = relationship(
        "WikiEntry",
        foreign_keys=[target_entry_id],
        back_populates="incoming_references"
    )
    context_chapter = relationship("Chapter", foreign_keys=[context_chapter_id])

    def __repr__(self):
        return f"<WikiCrossReference({self.reference_type}: {self.source_entry_id} -> {self.target_entry_id})>"

File: frontend/src/types/wiki.ts
/**
 * Wiki Types
 * Type definitions for the World Wiki system
 */

// ==================== Entry Types ====================

export type WikiEntryType =
  | 'character'
  | 'character_arc'
  | 'character_relationship'
  | 'location'
  | 'location_history'
  | 'magic_system'
  | 'world_rule'
  | 'technology'
  | 'culture'
  | 'religion'
  | 'faction'
  | 'artifact'
  | 'creature'
  | 'event'
  | 'timeline_fact'
  | 'theme';

export type WikiEntryStatus = 'draft' | 'published' | 'archived';

export type WikiChangeType = 'create' | 'update' | 'merge' | 'delete';

export type WikiChangeStatus = 'pending' | 'approved' | 'rejected';

export type WikiReferenceType =
  | 'mentions'
  | 'related_to'
  | 'part_of'
  | 'conflicts_with'
  | 'depends_on'
  | 'child_of'
  | 'ally_of'
  | 'enemy_of'
  | 'owns'
  | 'member_of'
  | 'located_in'
  // Culture-specific reference types
  | 'born_in'
  | 'exiled_from'
  | 'adopted_into'
  | 'leader_of'
  | 'rebel_against'
  | 'worships'
  | 'trades_with'
  | 'originated_in'
  | 'sacred_to'
  | 'resents';

// ==================== Wiki Entry ====================

export interface WikiEntry {
  id: string;
  world_id: string;
  entry_type: WikiEntryType;
  title: string;
  slug: string;
  content: string | null;
  structured_data: Record<string, any>;
  summary: string | null;
  image_url: string | null;
  parent_id: string | null;
  linked_entity_id: string | null;
  source_manuscripts: string[];
  source_chapters: Array<{
    manuscript_id: string;
    chapter_id: string;
    excerpt?: string;
  }>;
  status: WikiEntryStatus;
  confidence_score: number;
  tags: string[];
  aliases: string[];
  created_by: 'author' | 'ai' | 'codex_sync';
  created_at: string;
  updated_at: string | null;
}

export interface WikiEntryCreate {
  world_id: string;
  entry_type: WikiEntryType;
  title: string;
  content?: string;
  structured_data?: Record<string, any>;
  summary?: string;
  parent_id?: string;
  linked_entity_id?: string;
  tags?: string[];
  aliases?: string[];
}

export interface WikiEntryUpdate {
  title?: string;
  content?: string;
  structured_data?: Record<string, any>;
  summary?: string;
  status?: WikiEntryStatus;
  parent_id?: string;
  tags?: string[];
  aliases?: string[];
  image_url?: string;
}

// ==================== Cross References ====================

export interface WikiCrossReference {
  id: string;
  source_entry_id: string;
  target_entry_id: string;
  reference_type: WikiReferenceType;
  context: string | null;
  bidirectional: boolean;
  display_label: string | null;
  created_by: 'author' | 'ai';
  created_at: string;
}

export interface WikiCrossReferenceCreate {
  source_entry_id: string;
  target_entry_id: string;
  reference_type: WikiReferenceType;
  context?: string;
  bidirectional?: boolean;
}

// ==================== Change Queue ====================

export interface WikiChange {
  id: string;
  wiki_entry_id: string | null;
  world_id: string;
  change_type: WikiChangeType;
  field_changed: string | null;
  old_value: Record<string, any> | null;
  new_value: Record<string, any>;
  proposed_entry: WikiEntryCreate | null;
  reason: string | null;
  source_text: string | null;
  source_chapter_id: string | null;
  source_manuscript_id: string | null;
  confidence: number;
  status: WikiChangeStatus;
  priority: number;
  created_at: string;
  reviewed_at: string | null;
  reviewer_note: string | null;
  entry_title: string | null;
  entry_type: string | null;
}

// ==================== UI State ====================

export interface WikiFilters {
  entryType?: WikiEntryType;
  status?: WikiEntryStatus;
  parentId?: string;
  searchQuery?: string;
}

export interface WikiBrowserState {
  selectedEntry: WikiEntry | null;
  isEditing: boolean;
  filters: WikiFilters;
  viewMode: 'list' | 'grid' | 'tree';
}

// ==================== Entry Type Metadata ====================

export const WIKI_ENTRY_TYPE_INFO: Record<WikiEntryType, {
  label: string;
  icon: string;
  color: string;
  category: string;
}> = {
  character: { label: 'Character', icon: '👤', color: 'text-blue-600', category: 'Characters' },
  character_arc: { label: 'Character Arc', icon: '📈', color: 'text-blue-400', category: 'Characters' },
  character_relationship: { label: 'Relationship', icon: '🤝', color: 'text-blue-300', category: 'Characters' },
  location: { label: 'Location', icon: '🏠', color: 'text-green-600', category: 'Locations' },
  location_history: { label: 'Location History', icon: '📜', color: 'text-green-400', category: 'Locations' },
  magic_system: { label: 'Magic System', icon: '✨', color: 'text-purple-600', category: 'World Rules' },
  world_rule: { label: 'World Rule', icon: '📋', color: 'text-purple-400', category: 'World Rules' },
  technology: { label: 'Technology', icon: '⚙️', color: 'text-gray-600', category: 'World Rules' },
  culture: { label: 'Culture', icon: '🎭', color: 'text-orange-600', category: 'World Rules' },
  religion: { label: 'Religion', icon: '🙏', color: 'text-yellow-600', category: 'World Rules' },
  faction: { label: 'Faction', icon: '⚔️', color: 'text-red-600', category: 'Organizations' },
  artifact: { label: 'Artifact', icon: '💎', color: 'text-amber-600', category: 'Items' },
  creature: { label: 'Creature', icon: '🐉', color: 'text-emerald-600', category: 'Creatures' },
  event: { label: 'Historical Event', icon: '📅', color: 'text-indigo-600', category: 'Events' },
  timeline_fact: { label: 'Timeline Fact', icon: '⏰', color: 'text-cyan-600', category: 'Timeline' },
  theme: { label: 'Theme', icon: '💡', color: 'text-pink-600', category: 'Meta' },
};

export const WIKI_REFERENCE_TYPE_INFO: Record<WikiReferenceType, {
  label: string;
  reverseLabel: string;
}> = {
  mentions: { label: 'Mentions', reverseLabel: 'Mentioned by' },
  related_to: { label: 'Related to', reverseLabel: 'Related to' },
  part_of: { label: 'Part of', reverseLabel: 'Contains' },
  conflicts_with: { label: 'Conflicts with', reverseLabel: 'Conflicts with' },
  depends_on: { label: 'Depends on', reverseLabel: 'Required by' },
  child_of: { label: 'Child of', reverseLabel: 'Parent of' },
  ally_of: { label: 'Ally of', reverseLabel: 'Ally of' },
  enemy_of: { label: 'Enemy of', reverseLabel: 'Enemy of' },
  owns: { label: 'Owns', reverseLabel: 'Owned by' },
  member_of: { label: 'Member of', reverseLabel: 'Has member' },
  located_in: { label: 'Located in', reverseLabel: 'Contains' },
  // Culture-specific reference types
  born_in: { label: 'Born in', reverseLabel: 'Birthplace of' },
  exiled_from: { label: 'Exiled from', reverseLabel: 'Exiled' },
  adopted_into: { label: 'Adopted into', reverseLabel: 'Adopted' },
  leader_of: { label: 'Leader of', reverseLabel: 'Led by' },
  rebel_against: { label: 'Rebel against', reverseLabel: 'Opposed by' },
  worships: { label: 'Worships', reverseLabel: 'Worshipped by' },
  trades_with: { label: 'Trades with', reverseLabel: 'Trades with' },
  originated_in: { label: 'Originated in', reverseLabel: 'Origin of' },
  sacred_to: { label: 'Sacred to', reverseLabel: 'Holds sacred' },
  resents: { label: 'Resents', reverseLabel: 'Resented by' },
};

/** Culture-specific reference types for UI filtering */
export const CULTURE_REFERENCE_TYPES: WikiReferenceType[] = [
  'born_in',
  'exiled_from',
  'adopted_into',
  'leader_of',
  'rebel_against',
  'worships',
  'trades_with',
  'originated_in',
  'sacred_to',
  'resents',
  'member_of',
  'part_of',
  'located_in',
];

// ==================== API Response Types ====================

export interface WikiEntryReferences {
  outgoing: WikiCrossReference[];
  incoming: WikiCrossReference[];
}

export interface WikiSyncStats {
  total: number;
  created: number;
  updated: number;
  linked?: number;
  skipped?: number;
  errors: Array<{
    entity_id?: string;
    entity_name?: string;
    wiki_entry_id?: string;
    wiki_title?: string;
    error: string;
  }>;
}

export interface WikiValidationResult {
  violations: Array<{
    rule_id: string;
    rule_name: string;
    rule_type: string;
    severity: 'error' | 'warning' | 'info';
    message: string;
    text_excerpt: string;
  }>;
  count: number;
}
